File: Templates/_BusinessApp_/_BusinessApp_/Network/Base/BaseNetworkOperation.m
//
//  BaseNetworkOperation.m
//  MyInsurrance
//
//  Created by Gytenis Mikulėnas on 5/2/14.
//  Copyright (c) 2015 Gytenis Mikulėnas
//  https://github.com/GitTennis/SuccessFramework
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE. All rights reserved.
//

#import "BaseNetworkOperation.h"
#import "NSString+Validator.h"
#import "CustomAFJSONResponseSerializer.h"
#import "UserManager.h"

@implementation BaseNetworkOperation

@synthesize networkRequestObject = _networkRequestObject;
@synthesize context = _context;
@synthesize userManager = _userManager;

#pragma mark - Public -

// Forbit use of init
- (instancetype)init {
    
    NSAssert(NO, @"-init is not a valid initializer for the class %@. Use initWithNetworkRequestObject instead", NSStringFromClass([self class]));
    
    return nil;
}

- (instancetype)initWithNetworkRequestObject:(NetworkRequestObject *)networkRequestObject context:(id)context userManager:(id<UserManagerProtocol>)userManager {
    
    _context = context;
    _userManager = userManager;
    
    // Store request config
    _networkRequestObject = networkRequestObject;
    
    // Create network request url string
    NSString *urlString = [[self requestBaseUrl] stringByAppendingString:[self requestRelativeUrl]];
    
    // Add url params from the subclass if exist
    NSString *urlParamsString = [self requestUrlParams];
    if (urlParamsString.length > 0) {
        
        urlString = [[urlString stringByAppendingString:@"?"] stringByAppendingString:urlParamsString];
    }
    
    // Escape
    urlString = [urlString stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    
    // Create url request and proceed
    self = [self initWithUrlString:urlString networkRequestObject:networkRequestObject];
    if (self) {
        
    }
    return self;
}

- (id)initWithUrlString:(NSString *)urlString networkRequestObject:(NetworkRequestObject *)networkRequestObject  {
    
    if (![urlString isValidUrlString]) {
        
        DDLogError(@"[%@]: Invalid url: %@", [self class], urlString);
    }
    
    NSURL *url = [NSURL URLWithString:urlString];
    
    if (!url) {
        
        DDLogError(@"[%@]: Unable to create url from: %@", [self class], urlString);
    }
    
    // Create request
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
    
    // Set method
    [request setHTTPMethod:[self requestMethod]];
    
    // Add custom headers
    [self setHTTPHeadersWithRequest:request];
    
    // Add body
    [self setHTTPBodyWithRequest:request];
    
    //#ifdef DEBUG
    // For internal local https environment with self signed certificate
    //self.securityPolicy.allowInvalidCertificates = YES;
    //#endif
    
    self = [super initWithRequest:request];
    if (self) {
        
        // Add custom response serializer
        self.responseSerializer = [CustomAFJSONResponseSerializer serializer];
        
        // Nothing more is needed
    }
    return self;
}

- (void)performWithCallback:(Callback)callback {
    
    void(^successCallback)(AFHTTPRequestOperation *operation, id responseObject) = [self successWrappedCallbackWithPassedCallback:callback];
    void(^failCallback)(AFHTTPRequestOperation *operation, NSError *error) = [self failWrappedCallbackWithPassedCallback:callback];
    
    [self setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        
        successCallback(operation, responseObject);
        
    } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        
        failCallback(operation, error);
    }];
    
    [self start];
}

#pragma mark - Protected -

- (NSString *)requestBaseUrl {
    
    return _networkRequestObject.baseUrl;
}

- (NSString *)requestRelativeUrl {
    
    return _networkRequestObject.relativeUrl;
}

- (NSString *)requestMethod {
    
    return _networkRequestObject.method;
}

- (NSDictionary *)requestHeaders {
    
    NSString *token = _userManager.user.token;
    
    if (token.length > 0) {
        
        return @{@"Authorization" : [@"Bearer " stringByAppendingString:token]};
        
    } else {
        
        return nil;
    }
}

- (NSDictionary *)requestBodyParams {
    
    return nil;
}

- (NSString *)requestUrlParams {
    
    return nil;
}

- (void)setHTTPHeadersWithRequest:(NSMutableURLRequest *)urlRequest {
    
    // Add common headers
    [urlRequest setHTTPShouldHandleCookies:NO];
    [urlRequest setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [urlRequest setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    
    // Check for specific headers in subclass
    NSDictionary *customHeaders = [self requestHeaders];
    if (customHeaders) {
        
        for (NSString *header in [customHeaders allKeys]) {
            
            NSString *value = customHeaders[header];
            [urlRequest setValue:value forHTTPHeaderField:header];
        }
    }
}

- (void)setHTTPBodyWithRequest:(NSMutableURLRequest *)urlRequest {
    
    NSDictionary *bodyParams = [self requestBodyParams];
    
    if (bodyParams) {
        
        [urlRequest setHTTPBody:[NSJSONSerialization dataWithJSONObject:bodyParams options:NSJSONWritingPrettyPrinted error:nil]];
    }
}

- (void)handleReceivedDataWithSuccess:(BOOL)success result:(id)result error:(NSError *)error callback:(Callback)callback {
    
    NSAssert(NO, @"handleReceivedDataWithSuccess:result:error:callback is not implemented in class: %@", NSStringFromClass([self class]));
}

#pragma mark - Private -

- (void(^)(AFHTTPRequestOperation *operation, id responseObject))successWrappedCallbackWithPassedCallback:(Callback)callback {
    
    __weak __typeof(self)weakSelf = self;
    
    void (^successCallback)(AFHTTPRequestOperation *operation, id responseObject) = ^(AFHTTPRequestOperation *operation, id responseObject) {
        
        DDLogDebug(@"[%@]: success", [self class]);
        
        [weakSelf handleReceivedDataWithSuccess:YES result:responseObject error:nil callback:callback];
    };
    
    return successCallback;
}

- (void(^)(AFHTTPRequestOperation *operation, NSError *error))failWrappedCallbackWithPassedCallback:(Callback)callback {
    
    __weak __typeof(self)weakSelf = self;
    
    void (^failCallback)(AFHTTPRequestOperation *operation, NSError *error) = ^(AFHTTPRequestOperation *operation, NSError *error) {
        
        DDLogWarn(@"[%@]: failed with error: %@", [self class], error);
        
        if (error.code != kNetworkRequestErrorCanceledCode) {
            
            [[NSNotificationCenter defaultCenter] postNotificationName:kNetworkRequestErrorNotification object:nil userInfo:@{kNetworkRequestErrorNotificationUserInfoKey:error}];
        }
        
        [weakSelf handleReceivedDataWithSuccess:NO result:nil error:error callback:callback];
    };
    
    return failCallback;
}

@end

File: Templates/_BusinessAppSwift_/_BusinessAppSwift_/Network/BaseNetworkOperation.swift
//
//  BaseNetworkOperation.swift
//  _BusinessAppSwift_
//
//  Created by Gytenis Mikulenas on 22/10/16.
//  Copyright © 2016 Gytenis Mikulėnas
//  https://github.com/GitTennis/SuccessFramework
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE. All rights reserved.
//

import UIKit
import Alamofire

struct NetworkRequestError {
    
    static let notification = Notification.Name("NetworkRequestErrorNotification")
    static let notificationUserInfoErrorKey = "notificationUserInfoErrorKey"
    
    struct badData {
        
        static let code = 400
        static let name = localizedString(key:"NetworkRequestBadUserDataError")
    }
    
    struct unauthorized {
        
        static let code = 401
        static let name = localizedString(key:"NetworkRequestUnauthorizeError")
    }
    
    /*struct server {
     
     static let code = 500
     static let name = "NetworkRequestServerError"
     }*/
    
    struct canceled {
        
        static let code = -999
        static let name = localizedString(key:"NetworkRequestCanceledError")
    }
    
    struct timeout {
        
        static let code = -1001
        static let name = localizedString(key:"NetworkRequestTimeoutError")
    }
    
    struct isOffline {
        
        static let code = -1009
        static let name = localizedString(key:"NetworkRequestIsOfflineError")
    }
}

class BaseNetworkOperation: NetworkOperationProtocol {
    
    // MARK: NetworkOperationProtocol
    
    var networkRequest: NetworkRequestEntityProtocol
    var context: Any?
    var isSilent: Bool = false
    
    required init(networkRequest: NetworkRequestEntityProtocol, context: Any?, userManager: UserManagerProtocol, settingsManager: SettingsManagerProtocol) {
        
        self.context = context
        _userManager = userManager
        _settingsManager = settingsManager
        self.networkRequest = networkRequest
    }
    
    func perform(callback: @escaping Callback) {
        
        // Create network request url string
        var urlString: String = self.requestBaseUrl()
        if let requestRelativeUrl = self.requestRelativeUrl() {
            
            urlString = urlString + requestRelativeUrl
        }
        
        // Add url params from the subclass if exist
        if let urlParamsString: String = self.requestUrlParams() {
            
            urlString = urlString + "?" + urlParamsString
        }
        
        // Escape
        urlString = urlString.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed)!
        
        // Validate url
        /*if !urlString.isValidUrlString() {
         
         DDLogError(log: "Invalid url: " + urlString)
         }*/
        
        // Method
        let requestMethod: HTTPMethod = self.methodForName(name: self.requestMethod())
        
        // Headers
        let requestHeaders: Dictionary<String, String> = self.requestHeaders()
        
        // Body
        let requestBody: Dictionary<String, String>? = self.requestBodyParams()
        
        //#if DEBUG
        // For internal local https environment with self signed certificate
        //self.securityPolicy.allowInvalidCertificates = YES;
        //#endif
        
        Alamofire.request(urlString, method: requestMethod, parameters: requestBody, encoding: JSONEncoding.default, headers: requestHeaders)/*validate(statusCode: 200..<300)*/
            .validate(contentType: ["application/json"]).responseJSON { response in
                
                // Useful links for validation:
                // http://stackoverflow.com/a/34738244
                // http://stackoverflow.com/a/33102907
                
                let validationError = self.validate(response: response)
                
                // Immediatelly return error if custom validation fails
                if let error = validationError {
                    
                    let errorEntity = ErrorEntity.init(code: 0, message: "[" + className(object: self) + "]: failed with error: " + error.message)
                    DDLogError(log: errorEntity.message)
                    
                    self.handleResponse(success: false, result: nil, error: errorEntity, callback: callback)
                    
                    // Use Alamofire default validation
                } else {
                    
                    switch response.result {
                        
                    case .success(let value):
                        
                        DDLogDebug(log: "[" + className(object: self) + "]: success")
                        self.handleResponse(success: true, result: value, error: nil, callback: callback)
                        
                        break
                        
                    case .failure(let error):
                        
                        let errorEntity = ErrorEntity.init(code: 0, message: "[" + className(object: self) + "]: failed with error: " + error.localizedDescription)
                        DDLogError(log: errorEntity.message)
                        
                        self.handleResponse(success: false, result: nil, error: errorEntity, callback: callback)
                        
                        break
                    }
                }
        }
    }
    
    // MARK:
    // MARK: Protected
    // MARK:
    
    func requestBaseUrl() -> String {
        
        return networkRequest.baseUrl!
    }
    
    func requestRelativeUrl() -> String? {
        
        return networkRequest.relativeUrl
    }
    
    func requestMethod() -> String {
        
        return networkRequest.method
    }
    
    func requestHeaders() -> Dictionary<String, String> {
        
        var headers: Dictionary<String, String> = Dictionary()
        
        // Add common headers
        //[urlRequest setHTTPShouldHandleCookies:NO];
        
        headers["Accept"] = "application/json"
        headers["Content-Type"] = "application/json"
        
        let language: String = _settingsManager.language!
        let locale: String = Locale.current.identifier
        let localeComponents = locale.components(separatedBy: "_")
        let region: String = localeComponents.last!
        let acceptLanguage: String = language + "_" + region
        
        headers["Accept-Language"] = acceptLanguage
        
        // Add authorization if needed
        if let user = _userManager.user {
            
            if let token = user.token {
                
                headers["Authorization"] = "Bearer " + token
            }
        }
        
        return headers
    }
    
    func requestBodyParams() -> Dictionary<String, String>? {
        
        return nil
    }
    
    func requestUrlParams() -> String? {
        
        return nil
    }
    
    func handleResponse(success: Bool, result: Any?, error: ErrorEntity?, callback: Callback) {
        
        fatalError("handleResponse is not implemented in class: " + className(object: self))
    }
    
    // MARK:
    // MARK: Internal
    // MARK:
    
    internal var _userManager: UserManagerProtocol
    internal var _settingsManager: SettingsManagerProtocol
    
    func methodForName(name: String) -> HTTPMethod {
        
        var result: HTTPMethod = HTTPMethod.get
        
        switch name {
            
        case "OPTIONS":
            result = HTTPMethod.options
        case "GET":
            result = HTTPMethod.get
        case "HEAD":
            result = HTTPMethod.head
        case "POST":
            result = HTTPMethod.post
        case "PUT":
            result = HTTPMethod.put
        case "PATCH":
            result = HTTPMethod.patch
        case "DELETE":
            result = HTTPMethod.delete
        case "TRACE":
            result = HTTPMethod.trace
        case "CONNECT":
            result = HTTPMethod.connect
        default:
            result = HTTPMethod.get
        }
        
        return result
    }
    
    func validate(response: DataResponse<Any>)->ErrorEntity? {
        
        var result: ErrorEntity?
        
        if let httpStatusCode = response.response?.statusCode {
            
            var message : String = String(describing: response)
            
            let httpStatusCodeString: String = String(httpStatusCode)
            let index = httpStatusCodeString.index(httpStatusCodeString.startIndex, offsetBy: 1)
            let httpResponseStatusFirstDigit: String = httpStatusCodeString.substring(to: index)
            
            switch(httpResponseStatusFirstDigit) {
                
            case "2":
                
                // Everything is ok.
                break
                
            case "4":
                
                // Unauthorized
                if (httpStatusCodeString.isEqual(401)) {
                    
                    message = NetworkRequestError.unauthorized.name
                    
                    // 4XX bad data was passed
                } else {
                    
                    // ...
                }
                
                result = ErrorEntity.init(code: httpStatusCode, message: message)
                break
                
            case "5":
                
                result = ErrorEntity.init(code: httpStatusCode, message: message)
                break
                
            default:
                
                result = ErrorEntity.init(code: httpStatusCode, message: message)
                break
            }
        }
        
        return result
    }
}
